dali/util/odirect_file.cc

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2017-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <errno.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <cstdio>
+#include <cstring>
+#include <memory>
+#include <string>
+
+#include "dali/core/util.h"
+#include "dali/util/odirect_file.h"
+#include "dali/core/error_handling.h"
+
+namespace dali {
+
+static constexpr size_t kODirectAlignment = 4096;
+static constexpr size_t kODirectChunkSize = 2 << 20;  // 2M
+
+static size_t GetODirectEnvVals(const char *env_name, size_t min_value, size_t default_value,
+                                size_t alignment = 0) {
+  char *env = getenv(env_name);
+  int len = 0;
+  if (env && (len = strlen(env))) {
+    for (int i = 0; i < len; i++) {
+      bool valid = std::isdigit(env[i]) || (i == len - 1 && (env[i] == 'k' || env[i] == 'M'));
+      if (!valid) {
+        DALI_FAIL(make_string(
+          env_name, " must be a number, optionally followed by 'k' or 'M', got: ",
+          env));
+      }
+    }
+    size_t s = atoll(env);
+    if (env[len-1] == 'k')
+      s <<= 10;
+    else if (env[len-1] == 'M')
+      s <<= 20;
+    DALI_ENFORCE(is_pow2(s) || alignment != 0, make_string(env_name,
+                                                           " must be a power of two, got ", s));
+    DALI_ENFORCE(align_remainder(s, alignment) == 0 || alignment == 0,
```

Comment:
I know it's implemented with bitwise ops, but in general attempting to calculate a remainder with 0 divisor is ill advised.
```suggestion
    DALI_ENFORCE(alignment == 0 || align_remainder(s, alignment) == 0,
```